src/tallies/tally.cpp

```diff
@@ -503,49 +519,374 @@ Tally::init_triggers(pugi::xml_node node, int i_tally)
+        results_(i, j, RESULT_VALUE) = 0.0;
+        results_(i, j, RESULT_SUM) += val;
+        results_(i, j, RESULT_SUM_SQ) += val*val;
+      }
+    }
+  }
+}
+
 //==============================================================================
 // Non-member functions
 //==============================================================================
 
-adaptor_type<2> global_tallies()
+void read_tallies_xml()
 {
-  // Get pointer to global tallies
-  double* buffer;
-  openmc_global_tallies(&buffer);
+  // Check if tallies.xml exists. If not, just return since it is optional
+  std::string filename = settings::path_input + "tallies.xml";
+  if (!file_exists(filename)) return;
 
-  // Adapt into xtensor
-  std::array<size_t, 2> shape = {N_GLOBAL_TALLIES, 3};
-  std::size_t size {3*N_GLOBAL_TALLIES};
+  write_message("Reading tallies XML file...", 5);
 
-  return xt::adapt(buffer, size, xt::no_ownership(), shape);
-}
+  // Parse tallies.xml file
+  pugi::xml_document doc;
+  doc.load_file(filename.c_str());
+  pugi::xml_node root = doc.document_element();
 
-adaptor_type<3> tally_results(int idx)
-{
-  // Get pointer to tally results
-  double* results;
-  std::array<std::size_t, 3> shape;
-  // TODO: off-by-one
-  openmc_tally_results(idx+1, &results, shape.data());
+  // Check for <assume_separate> setting
+  if (check_for_node(root, "assume_separate")) {
+    settings::assume_separate = get_node_value_bool(root, "assume_separate");
+  }
+
+  // Check for user meshes and allocate
+  read_meshes(root);
+
+  // We only need the mesh info for plotting
+  if (settings::run_mode == RUN_MODE_PLOTTING) return;
+
+  // Read data for tally derivatives
+  read_tally_derivatives(root);
+
+  // ==========================================================================
+  // READ FILTER DATA
+
+  // Check for user filters and allocate
+  for (auto node_filt : root.children("filter")) {
+    // Copy filter id
+    if (!check_for_node(node_filt, "id")) {
+      fatal_error("Must specify id for filter in tally XML file.");
+    }
+    int filter_id = std::stoi(get_node_value(node_filt, "id"));
+
+    // Check to make sure 'id' hasn't been used
+    if (model::filter_map.find(filter_id) != model::filter_map.end()) {
+      fatal_error("Two or more filters use the same unique ID: "
+        + std::to_string(filter_id));
+    }
+
+    // Convert filter type to lower case
+    std::string s;
+    if (check_for_node(node_filt, "type")) {
+      s = get_node_value(node_filt, "type", true);
+    }
+
+    // Make sure bins have been set
+    if (s == "energy" || s == "energyout" || s == "mu" || s == "polar"
+      || s == "azimuthal" || s == "mesh" || s == "meshsurface" || s == "universe"
+      || s == "material" || s == "cell" || s == "distribcell" || s == "cellborn"
+      || s == "cellfrom" || s == "surface" || s == "delayedgroup") {
+      if (!check_for_node(node_filt, "bins")) {
+        fatal_error("Bins not set in filter " + std::to_string(filter_id));
```

Comment:
We should probably move this check into the `from_xml` methods of each filter.  It'll be a lot of duplication, but if we don't do it I think we're likely to forget to update this check with new filters.

Reply:
Yeah, I was also thinking that as I was rewriting this bit. I'll go ahead and move it.

Reply:
On second look, `get_node_value` (which is called in `from_xml` for each filter) already has a check for whether the node is missing and will call `fatal_error`, so this check is somewhat redundant. How do you feel about just removing it?

Reply:
Good point.  Yeah, we can just remove this check.